HeterogeneousCore/AlpakaInterface/interface/CachingAllocator.h

```diff
@@ -126,383 +102,56 @@ namespace cms::alpakatools {
-      } catch (std::runtime_error const& e) {
-        // the allocation attempt failed: free all cached blocks on the device and retry
-        if (debug_) {
-          std::ostringstream out;
-          out << "\t" << deviceType_ << " " << alpaka::getName(device_) << " failed to allocate " << block.bytes
-              << " bytes for queue " << block.queue->m_spQueueImpl.get()
-              << ", retrying after freeing cached allocations" << std::endl;
-          std::cout << out.str() << std::endl;
-        }
-        // TODO implement a method that frees only up to block.bytes bytes
-        freeAllCached();
-
-        // throw an exception if it fails again
-        block.buffer = allocateBuffer(block.bytes, *block.queue);
-      }
-
-      // create a new event associated to the "synchronisation device"
-      block.event = Event{block.device()};
-
-      {
-        std::scoped_lock lock(mutex_);
-        cachedBytes_.live += block.bytes;
-        cachedBytes_.requested += block.requested;
-        // TODO use std::move() ?
-        liveBlocks_[block.buffer->data()] = block;
-      }
-
-      if (debug_) {
-        std::ostringstream out;
-        out << "\t" << deviceType_ << " " << alpaka::getName(device_) << " allocated new block at "
-            << block.buffer->data() << " (" << block.bytes << " bytes associated with queue "
-            << block.queue->m_spQueueImpl.get() << ", event " << block.event->m_spEventImpl.get() << "." << std::endl;
-        std::cout << out.str() << std::endl;
-      }
-    }
-
-    void freeAllCached() {
-      std::scoped_lock lock(mutex_);
-
-      while (not cachedBlocks_.empty()) {
-        auto iBlock = cachedBlocks_.begin();
-        cachedBytes_.free -= iBlock->second.bytes;
-
-        if (debug_) {
-          std::ostringstream out;
-          out << "\t" << deviceType_ << " " << alpaka::getName(device_) << " freed " << iBlock->second.bytes
-              << " bytes.\n\t\t  " << (cachedBlocks_.size() - 1) << " available blocks cached (" << cachedBytes_.free
-              << " bytes), " << liveBlocks_.size() << " live blocks (" << cachedBytes_.live << " bytes) outstanding."
-              << std::endl;
-          std::cout << out.str() << std::endl;
-        }
-
-        cachedBlocks_.erase(iBlock);
-      }
-    }
-
-    // TODO replace with a tbb::concurrent_multimap ?
-    using CachedBlocks = std::multimap<unsigned int, BlockDescriptor>;  // ordered by the allocation bin
-    // TODO replace with a tbb::concurrent_map ?
-    using BusyBlocks = std::map<void*, BlockDescriptor>;  // ordered by the address of the allocated memory
+    Buffer allocateBuffer(size_t bytes, Queue const& queue);
 
-    inline static const std::string deviceType_ = alpaka::core::demangled<Device>;
+    void allocateNewBlock(BlockDescriptor& block);
+
+    void freeAllCached();
+
+    struct BlockList {
+      tbb::concurrent_queue<BlockDescriptor*> blocks_;
```

Comment:
Consider using a `tbb::concurren_queue<std::unique_ptr<BlockDescriptor>> blocks_`

Reply:
Thanks for the suggestion, I'm playing with it, and together with the other suggestions it does make ownership more clear.

Now I'm thinking if `tbb::concurrent_queue<BlockDescriptor>` could work better, since `BlockDescriptor` is small enough and is mostly made up of movable objects.

Reply:
Mhm, no, using the `BlockDescriptor` by value (even with moves) has a significant overhead, about 1-2% of the throughput for a GPU-only HLT job.